pageserver/src/layered_repository.rs

```diff
@@ -352,13 +354,35 @@ impl LayeredRepository {
         timelineid: ZTimelineId,
         tenantid: ZTenantId,
         data: &TimelineMetadata,
+        first_save: bool,
     ) -> Result<PathBuf> {
-        let path = conf.timeline_path(&timelineid, &tenantid).join("metadata");
-        let mut file = File::create(&path)?;
+        let timeline_path = conf.timeline_path(&timelineid, &tenantid);
+        let path = timeline_path.join("metadata");
+        // use OpenOptions to ensure file presence is consistent with first_save
+        let mut file = OpenOptions::new()
+            .write(true)
+            .create_new(first_save)
+            .open(&path)?;
 
         info!("saving metadata {}", path.display());
 
-        file.write_all(&TimelineMetadata::ser(data)?)?;
+        let mut metadata_bytes = TimelineMetadata::ser(data)?;
+        // Taken from PG_CONTROL_MAX_SAFE_SIZE
+        const MAX_SAFE_SIZE: usize = 512;
+        assert!(metadata_bytes.len() <= MAX_SAFE_SIZE);
+
+        metadata_bytes.resize(MAX_SAFE_SIZE, 0u8);
```

Comment:
I reworked this approach so it always writes the same # of bytes and doesn't truncate. Happy to switch to rename approach as well, let me know